beacon-chain/forkchoice/protoarray/forkchoice.go

```diff
@@ -0,0 +1,281 @@
+	_, span := trace.StartSpan(ctx, "protoArrayForkChoice.ProcessAttestation")
+	defer span.End()
+	f.votesLock.Lock()
+	defer f.votesLock.Unlock()
+
+	for _, index := range validatorIndices {
+		// Validator indices will grow the vote cache.
+		for index >= uint64(len(f.votes)) {
+			f.votes = append(f.votes, Vote{currentRoot: params.BeaconConfig().ZeroHash, nextRoot: params.BeaconConfig().ZeroHash})
+		}
+
+		// Newly allocated vote if the root fields are untouched.
+		newVote := f.votes[index].nextRoot == params.BeaconConfig().ZeroHash &&
+			f.votes[index].currentRoot == params.BeaconConfig().ZeroHash
+
+		// Vote gets updated if it's newly allocated or high target epoch.
+		if newVote || targetEpoch > f.votes[index].nextEpoch {
+			f.votes[index].nextEpoch = targetEpoch
+			f.votes[index].nextRoot = blockRoot
+		}
+	}
+
+	processedAttestationCount.Inc()
+}
+
+// ProcessBlock processes a new block by inserting it to the fork choice store.
+func (f *ForkChoice) ProcessBlock(
+	ctx context.Context,
+	slot types.Slot,
+	blockRoot, parentRoot [fieldparams.RootLength]byte,
+	justifiedEpoch, finalizedEpoch types.Epoch, optimistic bool,
+) error {
+	ctx, span := trace.StartSpan(ctx, "protoArrayForkChoice.ProcessBlock")
+	defer span.End()
+
+	return f.store.insert(ctx, slot, blockRoot, parentRoot, justifiedEpoch, finalizedEpoch, optimistic)
+}
+
+// Prune prunes the fork choice store with the new finalized root. The store is only pruned if the input
+// root is different than the current store finalized root, and the number of the store has met prune threshold.
+func (f *ForkChoice) Prune(ctx context.Context, finalizedRoot [32]byte) error {
+	return f.store.prune(ctx, finalizedRoot)
+}
+
+// HasNode returns true if the node exists in fork choice store,
+// false else wise.
+func (f *ForkChoice) HasNode(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	_, ok := f.store.nodeByRoot[root]
+	return ok
+}
+
+// HasParent returns true if the node parent exists in fork choice store,
+// false else wise.
+func (f *ForkChoice) HasParent(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	node, ok := f.store.nodeByRoot[root]
+	if !ok || node == nil {
+		return false
+	}
+
+	return node.parent != nil
+}
+
+// IsCanonical returns true if the given root is part of the canonical chain.
+func (f *ForkChoice) IsCanonical(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	node, ok := f.store.nodeByRoot[root]
+	if !ok {
+		return false
+	}
+
+	if node.bestDescendant == nil {
+		if f.store.headNode.bestDescendant == nil {
+			return node == f.store.headNode
+		}
+		return node == f.store.headNode.bestDescendant
+	}
+	if f.store.headNode.bestDescendant == nil {
+		return node.bestDescendant == f.store.headNode
+	}
+	return node.bestDescendant == f.store.headNode.bestDescendant
```

Comment:
By correctness `f.store.headNode.bestDescendant` should always be nil right?

It feels weird to cover the cases for `f.store.headNode.bestDescendant` is not nil. I rather not do it for unconsequence reasons, and we could be masking bugs here. I rather have it fails (ie return false)

Reply:
Resolving as per discussion:
https://discord.com/channels/476244492043812875/774005284188323870/943840351494492210